File: Solvers/mosek/10.0/tools/examples/python/opt_server_async.py
##
#  Copyright : Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#  File :      opt_server_async.py
#
#  Purpose :   Demonstrates how to use MOSEK OptServer
#              to solve optimization problem asynchronously
##
import mosek
import sys
import time

def streamprinter(msg):
    sys.stdout.write(msg)
    sys.stdout.flush()

if len(sys.argv) < 4:
    print("Missing argument, syntax is:")
    print("  opt-server-async inputfile host:port numpolls [cert]")
else:
    filename = sys.argv[1]
    addr = sys.argv[2]
    numpolls = int(sys.argv[3])
    token = None
    cert = None if len(sys.argv) < 5 else sys.argv[4]


    with mosek.Env() as env:

        with env.Task(0, 0) as task:

            print("reading task from file")
            task.readdata(filename)

            if cert is not None:
                task.putstrparam(mosek.sparam.remote_tls_cert_path,cert)

            print("Solve the problem remotely (async)")
            token = task.asyncoptimize(addr,"")

        print("Task token: %s" % token)

        with env.Task(0, 0) as task:

            task.readdata(filename)

            if cert is not None:
                task.putstrparam(mosek.sparam.remote_tls_cert_path,cert)
            task.set_Stream(mosek.streamtype.log, streamprinter)

            i = 0

            while i < numpolls:

                time.sleep(0.1)

                print("poll %d..." % i)
                respavailable, res, trm = task.asyncpoll(addr,
                                                         "",
                                                         token)

                print("done!")

                if respavailable:
                    print("solution available!")

                    respavailable, res, trm = task.asyncgetresult(addr,
                                                                  "",
                                                                  token)

                    task.solutionsummary(mosek.streamtype.log)
                    break

                i = i + 1

                if i == numpolls:
                    print("max number of polls reached, stopping host.")
                    task.asyncstop(addr,"", token)

File: Solvers/mosek/10.0/tools/examples/dotnet/opt_server_async.cs
/*
   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.

   File:       opt_server_async.cs

   Purpose :   Demonstrates how to use MOSEK OptServer
               to solve optimization problem asynchronously

*/
using System;
using System.Threading;

namespace mosek.example
{
  class msgclass : mosek.Stream
  {
    public override void streamCB (string msg)
    {
      Console.Write ("{0}", msg);
    }
  }

  public class opt_server_async
  {
    public static void Main (string[] args)
    {
      if (args.Length == 0) {
        Console.WriteLine ("Missing argument, syntax is:");
        Console.WriteLine ("  opt_server inputfile host:port numpolls [cert]");
      }
      else {

        string inputfile = args[0];
        string addr      = args[1];
        int    numpolls  = Convert.ToInt32(args[2]);
        String cert      = args.Length < 4 ? null : args[3];

        using (mosek.Env env = new mosek.Env())
        {

          string token;

          using (mosek.Task task = new mosek.Task(env))
          {
            task.readdata (inputfile);
            if (cert != null)
              task.putstrparam(sparam.remote_tls_cert_path,cert);
            token = task.asyncoptimize (addr,"");
          }

          using (mosek.Task task = new mosek.Task(env))
          {
            task.readdata (inputfile);
            if (cert != null)
              task.putstrparam(sparam.remote_tls_cert_path,cert);
            task.set_Stream (mosek.streamtype.log, new msgclass ());
            Console.WriteLine("Starting polling loop...");

            int i = 0;

            while ( true )
            {
              Thread.Sleep(500);
              Console.WriteLine("poll {0}...\n", i);

              mosek.rescode trm;
              mosek.rescode resp;

              bool respavailable = task.asyncpoll( addr,
                                                   "",
                                                   token,
                                                   out resp,
                                                   out trm);


              Console.WriteLine("polling done");

              if (respavailable)
              {
                Console.WriteLine("solution available!");

                task.asyncgetresult(addr,
                                    "",
                                    token,
                                    out resp,
                                    out trm);

                task.solutionsummary (mosek.streamtype.log);
                break;
              }

              if (i == numpolls)
              {
                Console.WriteLine("max num polls reached, stopping host.");
                task.asyncstop (addr,"", token);
                break;
              }
              i++;
            }

          }
        }
      }
    }
  }
}
